qt/search_panel.cpp

```diff
@@ -224,34 +239,82 @@ void SearchPanel::OnSearchTextChanged(QString const & str)
 
   ClearResults();
 
-  // search even with empty query
-  if (!normalized.isEmpty())
+  if (normalized.isEmpty())
   {
-    m_params.m_query = normalized.toUtf8().constData();
-    auto const timestamp = ++m_timestamp;
-    m_params.m_onResults = [this, timestamp](search::Results const & results,
-                                             std::vector<search::ProductInfo> const & productInfo) {
-      GetPlatform().RunTask(Platform::Thread::Gui,
-                            std::bind(&SearchPanel::OnSearchResults, this, timestamp, results));
+    m_pDrawWidget->GetFramework().GetSearchAPI().CancelAllSearches();
+
+    // hide X button
+    m_pClearButton->setVisible(false);
+    return;
+  }
+
+  bool started = false;
+  auto const timestamp = ++m_timestamp;
+
+  switch (m_mode)
+  {
+  case search::Mode::Everywhere:
+  {
+    m_everywhereParams.m_query = normalized.toUtf8().constData();
+    m_everywhereParams.m_onResults = [this, timestamp](
+                                         search::Results const & results,
+                                         std::vector<search::ProductInfo> const & productInfo) {
+      GetPlatform().RunTask(
+          Platform::Thread::Gui,
+          std::bind(&SearchPanel::OnEverywhereSearchResults, this, timestamp, results));
     };
-    m_params.m_timeout = search::SearchParams::kDefaultDesktopTimeout;
+    m_everywhereParams.m_timeout = search::SearchParams::kDefaultDesktopTimeout;
 
-    if (m_pDrawWidget->Search(m_params))
-    {
-      // show busy indicator
-      if (!m_pAnimationTimer->isActive())
-        m_pAnimationTimer->start(200);
+    started = m_pDrawWidget->GetFramework().GetSearchAPI().SearchEverywhere(m_everywhereParams);
 
-      m_pClearButton->setFlat(true);
-      m_pClearButton->setVisible(true);
-    }
+    m_viewportParams = {};
   }
-  else
+  break;
+
+  case search::Mode::Viewport:
   {
-    m_pDrawWidget->GetFramework().CancelSearch(search::Mode::Everywhere);
+    m_viewportParams.m_query = normalized.toUtf8().constData();
+    m_viewportParams.m_onCompleted = [this](search::Results const & results) {
+      GetPlatform().RunTask(Platform::Thread::Gui, [this, results] {
+        m_pDrawWidget->GetFramework().FillSearchResultsMarks(true /* clear */, results);
+        StopBusyIndicator();
+      });
+    };
+    m_viewportParams.m_timeout = search::SearchParams::kDefaultDesktopTimeout;
 
-    // hide X button
-    m_pClearButton->setVisible(false);
+    started = m_pDrawWidget->GetFramework().GetSearchAPI().SearchInViewport(m_viewportParams);
+
+    m_everywhereParams = {};
+  }
+  break;
+
+  default: started = false; break;
```

Comment:
without break

Reply:
or better delete it, started is false by default

Reply:
why?

Reply:
because we usually do not write such way (in one line)

Reply:
ah, it's clang-format
reformatted manually, done